### ISSUE TITLE
Question about TTT learning rate (eta)

### ISSUE THREAD

**Issue by Karami-m** [test-time-training/ttt-lm-jax#16]
Thanks for your great work and publishing its code.

I have a question regarding the learning rate. Equations (6) and (7) suggest that η should be a vector of size $mini\\_batch\\_size \times 1$ However, the code implements $\eta$ as a square matrix of size $mini\\_batch\\_size \times mini\\_batch\\_size$. This matrix is constructed  by multiplying the vector $\eta$ by  `token_idx` in the [[Jax version](https://github.com/test-time-training/ttt-lm-jax/blob/ac8cdc8a43b811afe23c27d7e82eed34a747b19c/ttt/models/ttt_layer.py#L223)](https://github.com/test-time-training/ttt-lm-jax/blob/ac8cdc8a43b811afe23c27d7e82eed34a747b19c/ttt/models/ttt_layer.py#L223) 
and `token_eta` in the [[PyTorch version](https://github.com/test-time-training/ttt-lm-pytorch/blob/cd831db10c8c9a0f6340f02da5613316a8a92b67/ttt.py#L796)](https://github.com/test-time-training/ttt-lm-pytorch/blob/cd831db10c8c9a0f6340f02da5613316a8a92b67/ttt.py#L796).

It appears that instead of applying the same vector $\eta$  across all tokens, a decayed version of $\eta$  is used for each subsequent token's computation. Could you elaborate on the role of this square matrix?  Also, have you experimented with using a simple vectorized $\eta$  as described in equation (7)?

**Comment by karan-dalal** [comment]
Sorry for the late response here.

You are correct. The eta is first computed token independent. We then scale the eta by the most recent token's position in the mini-batch. This implicitly averages the gradients of all tokens, as we do in mini-batch GD.

1st token in the mini-batch: `W0 = W0 - eta1 * G1 = W0 - (x1 @ vector) * G1`
2nd token = `W2 = W0 - 1/2 (eta1 * G1 + eta2 * G2) = W0 - 1/2 * eta1 * G1 - 1/2 * eta2 * G2`

The square matrix makes it easier to apply in our dual form. We didn't experiment without the mini-batch position scaling.
